### DISCUSSION TITLE
Leveling Z axis

### DISCUSSION THREAD

**Discussion by rubino25** [dz0ny/klipper-prusa-mk3s#7] [Q&A] [answered]
Hi, how to leveling the Z axis like the original firmware?

**Comment by dz0ny** [comment, marked as answer]
There is Z calibration macro https://github.com/dz0ny/klipper-prusa-mk3s/blob/main/macros.cfg#L168, it should show under macros section in UI.

**Reply by dev-zetta** [reply]
Actually, the z axis is not going to be aligned when you execute that macro. It's a dual Z printer, we need to move the Z axis to the endstops, with a lower driver current, until the steppers skips steps. Like stock firmware does.

Not sure if it's correct, but it "works":

[gcode_macro MECHANICAL_GANTRY_CALIBRATION]
gcode:
   {% set my_current = params.CURRENT|default(0.12)|float %} ; adjust crash current as required without fighting config.
    {% set oldcurrent = printer.configfile.settings["tmc2130 stepper_z"].run_current %} ; TODO: Find runtime current settings
    {% set oldhold = printer.configfile.settings["tmc2130 stepper_z"].hold_current %} 
    {% set x_max = printer.toolhead.axis_maximum.x %} 
    {% set y_max = printer.toolhead.axis_maximum.y %} 
    {% set z_max = printer.toolhead.axis_maximum.z %} 
    
    G28 ; todo query for homed
    G90 ; absolute
    G0 X{x_max / 2} Y{y_max / 2} F8000 ; TODO grab feedrate from config
    G0 Z{z_max} F720 ; go to the Z-max at speed 12*60
    SET_TMC_CURRENT STEPPER=stepper_z CURRENT={my_current}
    G4 P200 ; Probably not necessary, it is here just for sure
    FORCE_MOVE STEPPER=stepper_z DISTANCE=10 VELOCITY=6
    G4 P200 ; 
    FORCE_MOVE STEPPER=stepper_z DISTANCE=-4 VELOCITY=6
    G4 P200 ; same as the first one
    SET_TMC_CURRENT STEPPER=stepper_z CURRENT={oldcurrent} HOLDCURRENT={oldhold}
    G4 P200 ; same as the first one
    G28 Z

**Reply by dz0ny** [reply]
That's why the macro has ¨FORCE_MOVE STEPPER=stepper_z Distance=20 Velocity=10¨

**Reply by dev-zetta** [reply]
But why doesn't it work? I doesn't move Z axis to the endstop in first place btw.

**Reply by dev-zetta** [reply]
Btw, stepper_z has position_max: 210, but prusa.cfg overrides that to 200. That would explain that 1 cm gap. But you need to move it to 215 or 220 to get both sides aligned anyway.

**Reply by dz0ny** [reply]
and Distance=20 moves it past virtual end-stop, so to the 220. 

So in your case when you are duplicating the work you are basically saying, move to virtual endstop, force move 10 to 210, move again to virtual endstop, and then force-move a bit more. Could be that something is a bit different for your HW combo. So far no one else reported this issue nor I can duplicate it.

**Reply by confuzled-dev** [reply]
I also have a 1cm gap during this test. Shouldnt it buzz against the top surface to ensure that both z steppers are aligned?

**Reply by agravelot** [reply]
It's not an ideal solution, but i did overwrite max_position :

```
[stepper_z]
position_max: 230
```

**Reply by confuzled-dev** [reply]
> It's not an ideal solution, but i did overwrite max_position :
> 
> ```
> [stepper_z]
> position_max: 230
> ```
Does doing this cause a similar buzz to the stock fw for you? On my end, it went slightly higher but did not physically butt against the top of the machine as it did on stock. I can not get buzz akin to the stock performance out of the original code, the "position_max" modification, or the code posted in this discussion. 

Hoping for more details

**Reply by dev-zetta** [reply]
> > It's not an ideal solution, but i did overwrite max_position :
> > ```
> > [stepper_z]
> > position_max: 230
> > ```
> 
> Does doing this cause a similar buzz to the stock fw for you? On my end, it went slightly higher but did not physically butt against the top of the machine as it did on stock. I can not get buzz akin to the stock performance out of the original code, the "position_max" modification, or the code posted in this discussion.
> 
> Hoping for more details

Same here. I increased it to 240 I believe. Any value is okay, it will just try to go higher, but at lower current, so it won't damage the stepper motors

**Reply by agravelot** [reply]
Here is the stock prusa code for z calibration, i can't find anything related to lower current. Even there is a  `FORCE_HIGH_POWER_START` call. 
https://github.com/prusa3d/Prusa-Firmware/blob/82343184e7b76583ccb2f86c855e20d291aac8cb/Firmware/Marlin_main.cpp#L3188

**Reply by confuzled-dev** [reply]
I have increased my position max to 240 and it is still not making contact with the top surface. 

https://youtu.be/dZsMspOAkDw?t=104
This is what I am looking to accomplish. 
https://youtu.be/D0hNXL3ATqw
This is the Calibrate Z from the cfg, with [stepper_z] position_max: 240
https://youtu.be/aRhx-jQAH_M
This is Mechanical Ganty Calibration

Results:

Calibrate Z - Climbs but stops about 1cm short of max and does not buzz
Mechanical Gantry Calibration - Climbs but stops about 1cm short of max. This macro does buzz, but not against the top, so it falls slightly instead of bump leveling against the endstop.

**Reply by confuzled-dev** [reply]
Ok. I did a bunch of testing trying to get the proper buzz out of both Z Calibration and Mechanical Gantry Calibration. If you do not establish a higher z stop, I see no way to get a buzz with the default configuration

**Comment by dev-zetta** [comment]
I had to increase it too :)

On Thu, Apr 13, 2023 at 12:20 AM Antoine Gravelot ***@***.***>
wrote:

> It's not an ideal solution, but i did overwrite max_position :
>
> [stepper_z]
> position_max: 230
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dz0ny/klipper-prusa-mk3s/discussions/7#discussioncomment-5597752>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZBWB6STN7YTUL6HWBJO3DXA4TDHANCNFSM5YR4XV3Q>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Have a nice day,

Gabriel

**Reply by confuzled-dev** [reply]
this just tells it the new end stop is 10mm higher than it really is so it buzzes?
